File: Top_100_Liked_Questions/20._Valid_Parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        
        # using hash set and hash map for constant time check
        openingParen = set(['(', '{', '['])
        matchingParen = {
            ')': '(',
            '}': '{',
            ']': '['
        }
        
        for paren in s:
            if paren in openingParen:
                stack.append(paren)
            else:
                if not stack or stack.pop() != matchingParen[paren]:
                    return False
        
        return len(stack) == 0

File: Top_100_Liked_Questions/20._Valid_Parentheses/solution.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    
    const stack = [];
    const openingParen = {
        '(': 1,
        '{': 1,
        '[': 1
    }
    const matchingParen = {
        ')': '(',
        '}': '{',
        ']': '['
    }
    
    for (const paren of s) {
        if (openingParen[paren]) {
            // if it is an opening paren, push it to the stack
            stack.push(paren);
        } else {
            // if it is a closing paren, pop the stack and check if it is a pair
            if (!stack.length || matchingParen[paren] !== stack.pop()) {
                return false;
            }
        }
    }
    
    return stack.length === 0;
};